alf/algorithms/seditor_algorithm.py

```diff
@@ -0,0 +1,583 @@
+
+        return AlgStep(
+            output=ap_out.output, state=state, info={
+                'ap_out': ap_out,
+                **imgs
+            })
+
+    def rollout_step(self, inputs: TimeStep, state):
+        ap_out = self._predict_action(inputs, state, mode=Mode.rollout)
+        return AlgStep(output=ap_out.output, state=state, info=ap_out)
+
+    def train_step(self, inputs: TimeStep, state, rollout_info: SEditorInfo):
+        def _calc_entropy(dist, a):
+            return -dist_utils.compute_log_probability(dist, a)
+
+        ap_out = self._predict_action(
+            inputs, state, mode=Mode.train, rollout_info=rollout_info)
+
+        a_entropy = _calc_entropy(ap_out.a_dist, ap_out.a)
+        da_entropy = _calc_entropy(ap_out.da_dist, ap_out.da)
+
+        actor_loss = self._actor_train_step(inputs, ap_out, a_entropy)
+        da_loss, change_a_loss, a_l2 = self._d_actor_train_step(
+            inputs, ap_out, da_entropy)
+        actor_info = SEditorActorInfo(
+            a_loss=actor_loss,
+            da_loss=da_loss,
+            a_entropy=a_entropy,
+            da_entropy=da_entropy,
+            change_a_loss=change_a_loss,
+            a_l2=a_l2)
+
+        critic_info = self._critic_train_step(inputs, ap_out, rollout_info)
+
+        alpha_loss = self._alpha_train_step(a_entropy, da_entropy)
+
+        info = SEditorInfo(
+            reward=inputs.reward,
+            step_type=inputs.step_type,
+            discount=inputs.discount,
+            actor=actor_info,
+            critic=critic_info,
+            alpha=alpha_loss)
+
+        return AlgStep(output=ap_out.output[0], state=state, info=info)
+
+    def _compute_critics(self,
+                         critic_net,
+                         observation,
+                         action,
+                         replica_min=True):
+        """Compute Q(s,a)"""
+        observation = (observation, action)
+        critics, _ = critic_net(observation)  # [B, replicas * reward_dim]
+        critics = critics.reshape(  # [B, replicas, reward_dim]
+            -1, self._num_critic_replicas, *self._reward_spec.shape)
+        if replica_min:
+            if self.has_multidim_reward():
+                sign = self.reward_weights.sign()
+                critics = (critics * sign).min(dim=1)[0] * sign
+            else:
+                critics = critics.min(dim=1)[0]
+
+        return critics
+
+    def _actor_loss_fn(self, dqda, action):
+        loss = 0.5 * losses.element_wise_squared_loss(
+            (dqda + action).detach(), action)
+        return loss.sum(list(range(1, loss.ndim)))
+
+    def _actor_train_step(self, inputs, ap_out, a_entropy):
+        """Train UM policy according to Eq 7a, to maximize the utility Q.
+        """
+        alpha = self._log_alpha[0].exp().detach()
+
+        critics = self._compute_critics(self._critic_networks,
+                                        inputs.observation, ap_out.output[0])
```

Comment:
`ap_out.output[0], ap_out.output[1], ap_out.output[2]` makes the code hard to understand. It's better to use some meaning name other than 0, 1, 2.

Reply:
> `ap_out.output[0], ap_out.output[1], ap_out.output[2]` makes the code hard to understand. It's better to use some meaning name other than 0, 1, 2.

updated